notebooks/Numba_047_release_demo.ipynb

```diff
@@ -0,0 +1,559 @@
+   "source": [
+    "Dynamic function definition\n",
+    "=======================\n",
+    "\n",
+    "The 0.47.0 release adds a new capability to Numba, that of dynamic function generation. Essentially functions (closures) defined in a JIT decorated function can now \"escape\" the function they are defined in and be used as arguments in subsequent function calls. For example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# takes a function and calls it with argument arg, multiplies the result by 7\n",
+    "@njit\n",
+    "def consumer(function, arg):\n",
+    "    return function(arg) * 7\n",
+    "\n",
+    "_GLOBAL = 5\n",
+    "\n",
+    "@njit\n",
+    "def generator_func():\n",
+    "    c = 10\n",
+    "\n",
+    "    def escapee(x): # closure, 'a' is a local, 'c' is a freevar, '_GLOBAL' is global\n",
+    "        a = 9\n",
+    "        return x * c + a * _GLOBAL\n",
+    "\n",
+    "    # data argument for the consumer call\n",
+    "    x = np.arange(5)\n",
+    "\n",
+    "    # escapee function is passed to the consumer function along with its argument\n",
+    "    return consumer(escapee, x)\n",
+    "\n",
+    "generator_func()\n",
+    "    "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Support for `map`, `filter`, `reduce`\n",
+    "============================\n",
+    "\n",
+    "The ability to create dynamic functions lead to being able to write support for `map`, `filter` and `reduce`. This makes it possible to write more \"pythonic\" code in Numba :-)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import operator\n",
+    "from functools import reduce\n",
+    "from numba.typed import List\n",
+    "\n",
+    "@njit\n",
+    "def demo_map_filter_reduce():\n",
+    "\n",
+    "    # This will be used in map\n",
+    "    def mul_n(x, multiplier):\n",
+    "        return x * multiplier\n",
+    "    \n",
+    "    # This will be used in filter\n",
+    "    V = 20\n",
+    "    def greater_than_V(x):\n",
+    "        return x > V # captures V from freevars\n",
+    "    \n",
+    "    # this will be used in reduce\n",
+    "    reduce_lambda = lambda x, y: (x * 2) + y\n",
+    "\n",
+    "    a = [x ** 2 for x in range(10)]    \n",
+    "    n = len(a)           \n",
+    "    return reduce(reduce_lambda, filter(greater_than_V, map(mul_n, a, range(n))))\n",
+    "\n",
+    "demo_map_filter_reduce()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Support for `list.sort()`/`sorted` with key\n",
+    "=================================\n",
+    "\n",
+    "A further extension born from the ability to create dynamic functions was being able to support the ``key`` argument to `list.sort` and `sorted`, a quick demonstration:"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
This cell prints a large warning containing the DependinDeprecationWarnings to the notbook itself. This may be confusing for first time users. Maybe we could silence the warnings, or use a typed-list?

Reply:
typed list doesn't work as it can't handle the optional that appears as the type in sort, so it'll have to be a warning suppression, I'll open a ticket about the actual problem